datafusion-examples/examples/function_factory.rs

```diff
@@ -0,0 +1,232 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use datafusion::error::Result;
+use datafusion::execution::config::SessionConfig;
+use datafusion::execution::context::{FunctionFactory, RegisterFunction, SessionContext};
+use datafusion_common::tree_node::{Transformed, TreeNode};
+use datafusion_common::{exec_err, internal_err, DataFusionError};
+use datafusion_expr::simplify::ExprSimplifyResult;
+use datafusion_expr::simplify::SimplifyInfo;
+use datafusion_expr::{CreateFunction, Expr, ScalarUDF, ScalarUDFImpl, Signature};
+use std::result::Result as RResult;
+use std::sync::Arc;
+
+/// This example shows how to utilize [FunctionFactory] to implement simple
+/// SQL-macro like functions using a `CREATE FUNCTION` statement. The same
+/// functionality can support functions defined in any language or library.
+///
+/// Apart from [FunctionFactory], this example covers
+/// [ScalarUDFImpl::simplify()] which is often used at the same time, to replace
+/// a function call with another expression at rutime.
+///
+/// This example is rather simple and does not cover all cases required for a
+/// real implementation.
+#[tokio::main]
+async fn main() -> Result<()> {
+    // First we must configure the SessionContext with our function factory
+    let ctx = SessionContext::new()
+        // register custom function factory
+        .with_function_factory(Arc::new(CustomFunctionFactory::default()));
+
+    // With the function factory, we can now call `CREATE FUNCTION` SQL functions
+
+    // Let us register a function called f which takes a single argument and
+    // returns that value plus one
+    let sql = r#"
+    CREATE FUNCTION f1(BIGINT)
+        RETURNS BIGINT
+        RETURN $1 + 1
+    "#;
+
+    ctx.sql(sql).await?.show().await?;
+
+    // Now, let us register a function called f2  which takes two arguments and
+    // returns the first argument added to the result of calling f1 on that
+    // argument
+    let sql = r#"
+    CREATE FUNCTION f2(BIGINT, BIGINT)
+        RETURNS BIGINT
+        RETURN $1 + f1($2)
+    "#;
+
+    ctx.sql(sql).await?.show().await?;
+
+    // Invoke f2, and we expect to see 1 + (1 + 2) = 4
```

Comment:
I removed some of the other things that didn't really help teach people how to use this feature any more (they were more like ensuring all features got covered) to keep it shorter